### PR TITLE
Open in modal or new tab, keyboard shortcut

### PR DESCRIPTION
**Pull request by chr314**
Added new option to open google translate in modal in the page
Added keyboard shortcuts  #6

@itsecurityco i guess its a nice change for the version 4 :smiley: 
also i think you need to update description, screenshots in firefox add-ons page

text_modal
page_modal
settings_page

### CODE REVIEW

**Comment by itsecurityco** [comment]
Awesome, I love it! testing right now :)

**Comment by chr314** [comment]
hey @itsecurityco its weird it works on my machine(lol), i checked on chromium too
fedora 29, firefox 66.0.1

you get any error in console(of background script and the current tab page)?

File: manifest.json [hunk 1]
```diff
@@ -31,15 +31,38 @@
         },
         "default_popup": "src/popup/popup.html"
     },
+    "content_scripts": [
+        {
+            "matches": [
+                "http://*/*",
```
[the remark this answers is not in the document]
Comment (chr314) [reply manifest.json]: content_scripts run inside of pages
modal open inside of page

File: manifest.json [hunk 2]
```diff
@@ -31,15 +31,38 @@
...
+                "*://translate.googleusercontent.com/*",
+                "*://translate.googleusercontent.cn/*"
+            ],
+            "js": ["src/content/translate_frame.js"],
+            "run_at": "document_end",
+            "all_frames": true
+        }
+    ],
     "permissions": [
+        "<all_urls>",
```
[the remark this answers is not in the document]
Comment (chr314) [reply manifest.json]: because modal run inside of page this is necessary

[the remark this answers is not in the document]
Comment (chr314) [reply manifest.json]: i changed it so it runs only in http(s) urls

**Comment by chr314** [comment]
I'm also working on a new mode for page translation, injecting google translate widget

in-page-translate

**Comment by itsecurityco** [comment]
> hey @itsecurityco its weird it works on my machine(lol), i checked on chromium too
> fedora 29, firefox 66.0.1
> 
> you get any error in console(of background script and the current tab page)?

@chr314 the browser console don't have error but the modal is empty.

image

Also the cmd has this error

`bash
[firefox/index.js][debug] Firefox stderr: JavaScript error: moz-extension://fc283d93-a681-487c-82d7-d61c11f2dc59/src/content/modal.js, line 21: TypeError: this.element.attachShadow is not a function
`

image

**Comment by chr314** [comment]
ok i found the problem 
attachShadow is available since version 63 https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow#Browser_compatibility

current version of firefox is 66

**Comment by itsecurityco** [comment]
> ok i found the problem
> attachShadow is available since version 63 https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow#Browser_compatibility
> 
> current version of firefox is 66
> 
> try to update firefox
> 
> `sudo apt update`
> `sudo apt upgrade`

nice! any alternative to this function to previous version of Firefox? take in mine that not all users will have the last version of Firefox.

image

The users are small, but still, if we could found works on previous version will be great, or maybe make the modal disabled when the version is < 63

**Comment by chr314** [comment]
> > ok i found the problem
> > attachShadow is available since version 63 https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow#Browser_compatibility
> > current version of firefox is 66
> > try to update firefox
> > `sudo apt update`
> > `sudo apt upgrade`
> 
> nice! any alternative to this function to previous version of Firefox? take in mine that not all users will have the last version of Firefox.
> 
> image
> 
> The users are small, but still, if we could found works on previous version will be great, or maybe make the modal disabled when the version is < 63

yes i can to it using iframe with srcdoc but i need to change some things in modal class, probably today

**Comment by zstal** [comment]
Hey, I just wanted to ask for a keyboard shortcut option, really glad to see it being developed. Can you perhaps make it happen via the Firefox keyboard bindings that are not available? That way the user could set their own preference.

Moreover, I see this new dialog above and would like to know: does this mean the extension will need additional permissions? The reason I chose this extension for translation is because it doesn't need access to all data on the page, it justs hands the selected piece of text to Google. I'd prefer not to have Google injected into all pages because of this...

Thanks

**Comment by zstal** [comment]
Hey, I just wanted to ask for a keyboard shortcut option, really glad to see it being developed. Can you perhaps make it happen via the Firefox keyboard bindings that are now available? That way the user could set their own preference.

Moreover, I see this new dialog above and would like to know: does this mean the extension will need additional permissions? The reason I chose this extension for translation is because it doesn't need access to all data on the page, it just hands the selected piece of text to Google. I'd prefer not to have Google injected into all pages because of this...

Thanks

**Comment by chr314** [comment]
hey @zstal 

i will add custom shortcut to options

in options you can choose if it will be opened inside of the page as modal or it will opened in new tab as it is now

in the modal google does not have access to the page you are viewing because google translate is loaded in iframe and the iframe injected in the page only if you try to translate something

last month i didn't had much time to work on this, i hope i finish the update these days

thank you for your feedback

**Comment by Raziel-SSJ** [comment]
@chr314 Nice enhancement, but why this "modal" things ?
I do like the pop-up that come off with the toolbar icon 
Toolbar icon pop-up
 and I would have prefer it opened with a shortcut rather than the modal things.
Maybe another shortcut to add ? 🤗
Thanks for your work anyway and have a good day 😘

**Comment by chr314** [comment]
the new permission "tabs"  is used to get the url of the active tab for the shortcut

**Comment by chr314** [comment]
i changed the the default shortcuts because Alt is not supported on firefox < 63 https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/commands#Key_combinations

tip: it's bad idea to use your pc as root, you can create a sudo user

**Comment by itsecurityco** [comment]
> i changed the the default shortcuts because Alt is not supported on firefox < 63 https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/commands#Key_combinations
> 
> tip: it's bad idea to use your pc as root, you can create a sudo user


Yep, but it's a clean virtual machine, it come with only root user created by default, I don't have anything more  in there besides some extensions.

The modal looks really awesome btw... I just need to make a commits to use the language locales in the HTML files. Really awesome work man, thanks!

**Comment by chr314** [comment]
:smile: thanks

**Comment by mtantawy** [comment]
Hi @itsecurityco , i wrote you this review https://addons.mozilla.org/en-US/firefox/addon/to-google-translate/reviews/1422461/
In fact I have reviewed the code before even writing the review, and I trust you intend nothing malicious with this, just to make it clear my intentions with the review and this comment :)

I also appreciate the convenience that comes with this update and new version, really impressive work, thank you!
I however have to ask if there is a way to continue using the extension without these features? to be honest, one have heard a LOT about little extensions that were taken over and converted to do malicious actions, and that's my real concern, I am sure you understand given your bio :wink:

**Comment by chr314** [comment]
Hi @mtantawy 
the permissions are hard-coded in the `manifest.json` and the user can't change them

but you always can load manually the extension as zip

so if you want load manually the version before these changes you can download the older source code https://github.com/itsecurityco/to-google-translate/tree/f8ff1232508bdb6833c46b949c2a79fdfc6144ea

- download the source code
```
git clone https://github.com/itsecurityco/to-google-translate.git

cd to-google-translate

git reset --hard f8ff1232508bdb6833c46b949c2a79fdfc6144ea
```
- build
```
make prepare

web-ext build (or zip the source code manually)
```

i created the older version zip if you want to use it [to_google_translate-3.9.zip](https://github.com/itsecurityco/to-google-translate/files/3646699/to_google_translate-3.9.zip)

- load the extension
- - go to `about:config` set `xpinstall.signatures.required` to `false` so you can load unsigned extensions
- - go to `about:addons` click the settings button on top right, select `Install Add-On From File...` and select the zip file

**Comment by itsecurityco** [comment]
> Hi @itsecurityco , i wrote you this review https://addons.mozilla.org/en-US/firefox/addon/to-google-translate/reviews/1422461/
> In fact I have reviewed the code before even writing the review, and I trust you intend nothing malicious with this, just to make it clear my intentions with the review and this comment :)
> 
> I also appreciate the convenience that comes with this update and new version, really impressive work, thank you!
> I however have to ask if there is a way to continue using the extension without these features? to be honest, one have heard a LOT about little extensions that were taken over and converted to do malicious actions, and that's my real concern, I am sure you understand given your bio 😉

Hello @mtantawy ,

I understand your concerns, but this usually happens form I understand when developers sell the extension to third companies, and this is not happening. Also Mozilla Firefox made a code review every new version. If you want still use the previous version you can follow @chr314 guide.

Cheers!

**Comment by mtantawy** [comment]
Thanks @chr314 for the steps, i have not updated the extension in Firefox so i still have the last version.
@itsecurityco, it doesn't have to be only that a developer sold the extension, it could be that a developer got hacked or something, if that happens (extension hacked) and the extension already has these permissions then users won't know if the next update is malicious or no, but if that happens (extension hacked) and it had no elevated permissions, then when users are requested to accept the elevated permissions they'll -hopefully- be alarmed, and raise the concern as being done in reviews and my comment here.
I am also sure that while code review by Mozilla is a step forward, it is not as thorough as one can imagine, they probably don't inspect js libs included for example.

It looks like i am not the only one raising the concern, judging by the reviews.
At this point, I'll have to stop getting updates for the extension unless this is reverted.

**Comment by Raziel-SSJ** [comment]
> It looks like i am not the only one raising the concern, judging by the reviews.
> At this point, I'll have to stop getting updates for the extension unless this is reverted.

You are not the only one and I totally agree.

Since you just could do the way I've mentioned in my post above, I'll not update.
I chose this extension for these low permissions needs and I rather prefer to support you financially than accept an extension that spies my navigation.

Regards.

**Comment by itsecurityco** [comment]
> > It looks like i am not the only one raising the concern, judging by the reviews.
> > At this point, I'll have to stop getting updates for the extension unless this is reverted.
> 
> You are not the only one and I totally agree.
> 
> Since you just could do the way I've mentioned in my post above, I'll not update.
> I chose this extension for these low permissions needs and I rather prefer to support you financially than accept an extension that spies my navigation.
> 
> Regards.

Nobody is spying anything, I don't understand the need to misinform and cause panic. The only fact is that in order to allow new tabs open in modal the manifest file was upgraded with tabs permissions. You also can choose for not update the add-on.

**Comment by Raziel-SSJ** [comment]
> Nobody is spying anything, I don't understand the need to misinform and cause panic. The only fact is that in order to allow new tabs open in modal the manifest file was upgraded with tabs permissions. You also can choose for not update the add-on.

Your are serious when saying "Nobody is spying anything" ?!
This is the actual disease of the web.

> "Access your data for all websites"
The extension can read the content of any web page you visit as well as data you enter into those web pages, such as usernames and passwords.

So stop hiding the potential spying on our activity and even more in fact.
How to know if there is no data mining in this by not being a developer ??
It is not because you are honest that you will not pay the price of developers with bad ethics whose the web is filled.

We are 2 to try to warn you that you are ruining the strong point this extension and, as mentioned above, adding only 2 shortcuts would have been enough.
But if you do not want to hear it... You are the boss !!

**Comment by chr314** [comment]
Hi @Raziel-SSJ 

> Are you serious when saying "Nobody is spying anything" ?!
> This is the actual disease of the web.

I'm pretty sure @itsecurityco is talking about this extension and not the entire web 


you or anybody else can always check the source code here on github or while it is loaded on the firefox, the running code is not minified or obfuscated 

i.e. `about:devtools-toolbox?type=extension&id=jid1-93WyvpgvxzGATw@jetpack` here you can check part which is running in the background, you can see the source code, network request etc


if you can't trust this version, in my previous comment i explained how you can build from the source code and load it manually  https://github.com/itsecurityco/to-google-translate/pull/46#issuecomment-534499626


before this a lot of complaints was why it's opening in new tab
i guess it's impossible to please everyone :see_no_evil: :hear_no_evil: :speak_no_evil: 


these permissions allow us to develop more features and help more people

**Comment by chr314** [comment]
working on less permissions #57

**Pull request merged**